### PR TITLE
dev-libs/re2: add 2023.08.01, mulitlibify abseil-cpp, add abseil-cpp-20230125.3

### PR DESCRIPTION
**Pull request by hhoffstaette**
re2 started to depend on abseil-cpp. As it is multilib-enabled, this PR also multilib-ifies the two most recent abseil-cpp versions.

Build- & runtime tested with grpc-1.52.1 and dnsdist (+regex).

This is a bit of a prelude to https://github.com/gentoo/gentoo/pull/31893 in order to reduce the scope and actually understand the abseil issue.

### CODE REVIEW

**Comment by hhoffstaette** [comment]
This is partially based on https://github.com/gentoo/gentoo/pull/31388 which was closed without conclusion, probably because chromium will start to use the bundled re2.
This leaves the question whether an updated commit like https://github.com/gentoo/gentoo/pull/31388/commits/fb2540ed5ff1d0f95f4171950dc14362e17e7347 should still be applied to qtwebengine, esp. 5.15.x.

**Comment by hhoffstaette** [comment]
@thesamesam as discussed on IRC. Older abseil releases than 2022.06 can probably be purged.

**Comment by stha09** [comment]
I should have mentioned that using bundled re2 in other packages is probably the best way to go. The library changes ABI with the C++ standard, e.g. you have absl::string_view symbols in C++14 mode and std::string_view symbols in C++17 mode. Also Chromium (and probably qtwebengine:6) is compiled in C++20 mode with bundled abseil, but you link it with something that is compiled in C++14 or C++17 mode.

**Comment by hhoffstaette** [comment]
Thanks for your comment, but somehow this does not really help us move forward.

The final conclusion here seems to be that we must abandon both abseil and re2 packages as libraries and only ever build+link them into any final executables, because that is where the C++ ABI is decided.

> I should have mentioned that using bundled re2 in other packages is probably the best way to go.

Not everything bundles re2 - e.g. dnsdist uses re2 as optional feature and simply links against it when selected. It runs fine.

> The library changes ABI with the C++ standard, e.g. you have absl::string_view symbols in C++14 mode and std::string_view symbols in C++17 mode.

But this is an abseil problem (and horrible misfeature, obviously), not directly re2's fault (its only fault is to rely on abseil). re2 only requires a minimum of C++14, but should also work when built as C++17. If necessary we should enforce C++17 (like the latest bseil-cppp ebuild does!) or at least always set `ABSL_USES_STD_STRING_VIEW`, _regardles of C++ standard_.
[The default option](https://github.com/abseil/abseil-cpp/blob/master/absl/base/options.h#L127) is of course the wrongest choice possible: compile-time-detection, which is why we are here.

> Also Chromium (and probably qtwebengine:6) is compiled in C++20 mode with bundled abseil, but you link it with something that is compiled in C++14 or C++17 mode.

Yes, both chromium and qtwebengine need to stop linking against re2 and use their bundled version. I'm not _entirely_ sure if that would really help though - I have a suspicion that this would only set us up for more complications later down the road. 

So what now? The purpose of _this_ PR is not to fix the world, but to have a small increment for abseil (multilib) required for re2 update. Not chromium, not grpc.

**Comment by hhoffstaette** [comment]
New comment because I finally understand what's going on.

Both ebuilds set -DCMAKE_CXX_STANDARD=17 - reasonable.

abseil-cpp-20220623.1 creates options.h with `ABSL_USE_STD_STRING_VIEW`=2
abseil-cpp-20230125.2 creates options.h with `ABSL_USE_STD_STRING_VIEW`=**1**

The ABI break is new in the 2023.* versions, probably because all this got too dumb even for Google.

However what i don't understand is why that is apparently such an insurmountable problem, because once we force abseil system-wide to **either** 0 (internal impl.) or 1 (alias to std::string_view) we should be good.
Yes, we would need a rebuild, but that seems to be the smallest problem.
Even packages that depend on an older version would be OK once we start forcing `ABSL_USE_STD_STRING_VIEW`=**1**.

Fundamentally the problem we have right now is **gone** once we have updated to abseil >= 2023.xx and always delegate to std::string_view.

**Comment by hhoffstaette** [comment]
So what do we think of the following snippet for the 20220623 version:

```
    # force use of std::* implementations for C++17
    sed -i -r -e 's/ABSL_OPTION_USE_STD_(.*) 2/ABSL_OPTION_USE_STD_\1 1/g' \
        absl/base/options.h || die
```

This is wht we should have been doing all along and creates an identical configuration to the 2023xxyy default when C++17 is enabled:
```
$grep STD_ /tmp/portage/dev-cpp/abseil-cpp-20220623.1-r1/work/abseil-cpp-20220623.1/absl/base/options.h
// ABSL_OPTION_USE_STD_ANY
// absl::any is a typedef of std::any, use the feature macro ABSL_USES_STD_ANY.
#define ABSL_OPTION_USE_STD_ANY 1
// ABSL_OPTION_USE_STD_OPTIONAL
// ABSL_USES_STD_OPTIONAL.
#define ABSL_OPTION_USE_STD_OPTIONAL 1
// ABSL_OPTION_USE_STD_STRING_VIEW
// ABSL_USES_STD_STRING_VIEW.
#define ABSL_OPTION_USE_STD_STRING_VIEW 1
// ABSL_OPTION_USE_STD_VARIANT
// ABSL_USES_STD_VARIANT.
#define ABSL_OPTION_USE_STD_VARIANT 1
```

**Comment by hhoffstaette** [comment]
I included the 20230125.3 update from [Sam's PR](https://github.com/gentoo/gentoo/pull/31893) and fixed the gtest dependency to be optional and actually work as expected.

It currently fails one test though:
```
99% tests passed, 1 tests failed out of 207

Total Test time (real) = 100.87 sec

The following tests FAILED:
	191 - absl_per_thread_sem_test (Failed)
```

**Comment by hhoffstaette** [comment]
Cherry on top: with the abseil-2022xxyy restriction locally removed from grpc-1.52-r1 and allowing any version, I updated to abseil-20230125.3, rebuilt re2/grpc/sysdig/libreoffice and all worked fine.

File: dev-cpp/abseil-cpp/abseil-cpp-20220623.1-r1.ebuild [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+	# un-hardcode abseil compiler flags
+	sed -i \
+		-e '/"-maes",/d' \
+		-e '/"-msse4.1",/d' \
+		-e '/"-mfpu=neon"/d' \
+		-e '/"-march=armv8-a+crypto"/d' \
+		absl/copts/copts.py || die
+
+	# force use of std::* implementations for C++17
```
Comment (thesamesam) [inline dev-cpp/abseil-cpp/abseil-cpp-20220623.1-r1.ebuild:50]: This merits a larger comment, reflecting the discussion in the PR. Please also do that in the commit message too.

**Comment by thesamesam** [comment]
> The change to force C++17 ABI (unconditional delegation to std:: classes) might be contentious, but it is the default from 2023xxyy onwards and I have not found a better way. The downside - and it is inevitable - is that all consumers of re2/grpc/abseil must be built with at least C++17. This sounds worse than it is, as that is the default level in gcc-13 and clang-16.
> 
> All my systems currently run abseil-20220623-r1 with forced C++17 mode. This caused a few rebuilds due to revbump. Other abseil-cpp consumers are:
> 
>     * dnsdist: via dynamically linked re2-2023.08.01 - no problems
> 
>     * libreoffice: already using `-std=c++17` and compile-time only, so no problems
> 
>     * sysdig _(obviously)_: it has a hardcoded `-std=c++0x` in its build, which of course fails when building a gRPC module, where `absl::string_view` is now declared to be `std::string_view` - which does not exist in C++11. I already fixed it up locally by forcing `-std=c++17` and it now builds & works fine as well.

Is this an exhaustive list, or just the ones you already had installed? 

Could you check the others if it's not-exhaustive, and also include the sysdig fix in this PR?

**Comment by hhoffstaette** [comment]
> Is this an exhaustive list, or just the ones you already had installed?

Just my systems (server & desktop).

> Could you check the others if it's not-exhaustive, and also include the sysdig fix in this PR?

The sysdig fix is easy. What do you mean by "checking the others"? I did a grep through @world and basically all ebuilds using older abseil are enforcing USE="cxx17", which is gone because we do that unconditionally now.

**Comment by thesamesam** [comment]
> > Is this an exhaustive list, or just the ones you already had installed?
> 
> Just my systems (server & desktop).
> 
> > Could you check the others if it's not-exhaustive, and also include the sysdig fix in this PR?
> 
> The sysdig fix is easy. What do you mean by "checking the others"? I did a grep through @world and basically all ebuilds using older abseil are enforcing USE="cxx17", which is gone because we do that unconditionally now.

If non-exhaustive, then check anything else depending on abseil. But you've answered that now I think.

**Comment by hhoffstaette** [comment]
> If non-exhaustive, then check anything else depending on abseil. But you've answered that now I think.

Understood. The real problem is anything depending on re2 or grpc, i.e. depending on abseil transitively.

File: dev-cpp/abseil-cpp/abseil-cpp-20220623.1-r1.ebuild (hunk 1 above)
Comment (hhoffstaette) [reply dev-cpp/abseil-cpp/abseil-cpp-20220623.1-r1.ebuild]: Long blurb added in ebuild and commit message. Hope it's clear enough.

**Comment by hhoffstaette** [comment]
Fwiw I expect there to be *some* fallout, simply because that's the price to pay for not using abseil properly to begin with. I was also very surprised to find the options.h stuff. 😕

**Comment by thesamesam** [comment]
@stha09 Happy with it now?

**Comment by thesamesam** [comment]
Looks like Chromium is using bundled re2 already, so we just need to update qtwebengine. Once that's done in this PR, I think we're good to go. Just then keep an eye on bugs.

**Comment by hhoffstaette** [comment]
> Looks like Chromium is using bundled re2 already, so we just need to update qtwebengine. Once that's done in this PR, I think we're good to go. Just then keep an eye on bugs.

Not sure what we have to do here as qtwebengine only requires "any re2", so a simple rebuild after the re2 update should suffice for ~arch? I don't see how this affects stable unless people mix keywords.
Also I won't be around over the weekend after today afternoon.

**Comment by thesamesam** [comment]
> > Looks like Chromium is using bundled re2 already, so we just need to update qtwebengine. Once that's done in this PR, I think we're good to go. Just then keep an eye on bugs.
> 
> Not sure what we have to do here as qtwebengine only requires "any re2", so a simple rebuild after the re2 update should suffice for ~arch? Ths should happen automagically since new re2 bumps its soname slot. I don't see how this affects stable unless people mix keywords. Unfortunately I won't be around over the weekend after today afternoon.

Aren't we now back to https://github.com/gentoo/gentoo/pull/31388#issuecomment-1587187251? If I'm missing something (and I might be), please say!

**Comment by stha09** [comment]
Again, qtwebengine is compiled in C+11 mode. The codebase is not ready for C++14 or even C++17.
Chromium is compiled in C++20 mode and mixing C++ standards already created strange runtime issues (Chromium used C++14 while Skia used C++17).
Pinning C++17 in abseil is in my opinion the wrong approach and will just introduce runtime issues. Why not use the bundled re2 in the consumers? Linking all the abseil shared libraries to re2 (I think around 10) also increases the load time drastically.

**Comment by hhoffstaette** [comment]
> Aren't we now back to [#31388 (comment)](https://github.com/gentoo/gentoo/pull/31388#issuecomment-1587187251)? If I'm missing something (and I might be), please say!

No, you are right. If qtwebengine cannot be built with C++17 then we're either screwed forever, must force an update to a version that allows us to do so, or make qtwebengine use a bundled re2 (possibly a version without abseil) just like chromum. However that is primarily an upstream problem. We are only in this situation because we insist on unbundling by default, and have to fight the rest of the world who don't want that.

This brings us back to a possible fallback plan: using abseil in non-delegation mode (0), i.e. always using the abseil implementations. I don't know which or how many ABI problems *that* would cause when abseil is built & configured with mode 0, possibly always using abseil's lowest common requirement (C++11 or -14 instead of forcing -17? ie. make it transitively less demanding, not more) and a consumer built with C++14 just links to absl classes, which should be fine AFAIK.
Maybe all consecutive problems will then disappear in a puff of smoke, that would be nice.

Using 0  would isolate us from the ABI - until Google ever decide to get rid of their pre-C++14/17 wrapper types.

Alternatively we just give up and stay with the last non-abseil using re2 forever. Until of course some package requires a newer version, in which case we end up here again, or with conflicts.

**Comment by hhoffstaette** [comment]
>  Pinning C++17 in abseil is in my opinion the wrong approach and will just introduce runtime issues. Why not use the bundled re2 in the consumers? Linking all the abseil shared libraries to re2 (I think around 10) also increases the load time drastically.

I basically just wrote the same thing and would love nothign more that for both abseil and re2 to become header-only packages.

**Comment by thesamesam** [comment]
Yeah, I think that's fair. qtwebengine can be switched to a bundled copy. What do we do about things without bundled copies? I don't think we gain anything from unbundling re2/abseil and I think if anything it makes life far worse, so I don't think this is a problem for us to do.

Is there anything significant without a bundled copy even available at all?

**Comment by hhoffstaette** [comment]
> Is there anything significant without a bundled copy even available at all?

I don't know exactly how CMake wants to "find a package" but at least dnsdist has an optional re2 dependency (I think [here](https://github.com/PowerDNS/pdns/blob/master/pdns/dnsdistdist/configure.ac#L218) and just links to it. Offhand I wouldn't know how to fix that.

**Comment by thesamesam** [comment]
> > Is there anything significant without a bundled copy even available at all?
> 
> I don't know exactly how configure or cmake wants to "find a package" but at least dnsdist has an optional re2 dependency (I think [here](https://github.com/PowerDNS/pdns/blob/master/pdns/dnsdistdist/configure.ac?rgh-link-date=2023-08-18T08%3A12%3A53Z#L218)) and just links to it. Offhand I wouldn't know how to fix that.
> 
> How would the packaging for re2 & abseil have to change (i.e. their ebuilds)? Would they still exist as ebuilds?

We'll have to keep them, I think, for the handful of things where they either hard-require it without a bundled copy or where they disable functionality we need.

But we'll have to try keep that small number of packages in sync with the C++ std version I guess.

**Comment by hhoffstaette** [comment]
> We can certainly choose to isolate ourselves from further ABI changes by building abseil in whatever lowest common -std it wants to have and never delegate (i.e. mode 0 as explained above). Right now that seems to be c++14.

As a quick experiment over my lunch break I rebuilt abseil with std=c++14, which disables delegation and creates options.h with "0" configuration, i.e. using abseil's implementations.
As expected this worked fine - all rebuilds succeeded and work, regardless of their own required/desired C++ version.

dnsdist/libreoffice/sysdig+fix build with c++17, grpc only requires c++14.

When qtwebkit uses a bundled re2, with or without abseil, this would provide the least bumy way forward.

Unless anyone can make a good argument why we should **not** build abseil with c++14 instead of 17 I'll add these changes hopefully Sunday evening.

**Comment by hhoffstaette** [comment]
> horrible conflicts for lo-bin users.

I took a look at this and have no idea why libreoffice-bin:
- wants to runtime-depend on abseil, as it has **no** such runtime dependencies anywhere, at least according to ldd.
- depends on both a fixed and an unspecified version

Neither does my compiled version:
`$find /usr/lib64/libreoffice -type f | xargs ldd 2>/dev/null | grep absl`
-> nothing!

So whatever is ailing LO is probably .. wrong?

**Comment by hhoffstaette** [comment]
I am cancelling this particular apocalypse. It makes no sense to try and _increase_ our exposure to abseil complications by bumping re2 now; the only sustainable way forward is to start doing it backwards, by:
- continuing to use the existing non-abseil-exposing re2 where necessary (unbundled); it's functional, stable and works well
- start re-bundling abseil wherever it is used 
- eventually re-bundle re2 as well, wherever necessary.

The one big potential headache would be grpc. Luckily, it seems that - unlike re2 - the grpc++ headers do **not** leak abseil types, so the abseil version used for building grpc does not matter immediately, and we can keep using a shared grpc.

**Comment by andreili** [comment]
@hhoffstaette 
I can't build or-tools library (https://github.com/google/or-tools) with this fixes (force use C++14), see issue https://github.com/google/or-tools/issues/3939

**Pull request closed without merging**